### PR TITLE
[WIP][V1][Spec Decode] EAGLE tree-attention

### PR DESCRIPTION
**Pull request by wwl2755**
As mentioned in #15901, currently we only support top-1 selection from the candidates from the EAGLE model (we call it chain-draft), and in EAGLE and EAGLE-2, both are claim select top-k tokens from each forward pass can benefit the acceptance rate, so we want to support it (we call it tree-draft). 

As this would be a big change, I would like to work on a WIP PR and would be appreciated to receive any comments/suggestions/discussion during implementation. 

Design Doc: https://docs.google.com/document/d/1mMoSicPPMMzaE_T5Zk2SnTderw1OXRUs2T16JxfVGCQ/edit?usp=sharing

cc: @LiuXiaoxuanPKU @WoosukKwon 

- [x] Construct tree structure
- [x] Select top-k instead of top-1 from the logits from targer model
- [x] Selection in Level-0 and level-1
- [x] Selection in Level-2 & 2+
- [x] Node/path selection after expansion
- [ ] Attention metadata & attention mask, customized ROPE embedding
- Now the nodes without ancestor-relation could see each other (which is incorrect)
- [ ] Rejection logic
- [ ] KV cache & CUDA graph (future PRs)
- [ ] E2e and unit tests

### CODE REVIEW

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -18,6 +18,193 @@
 PADDING_SLOT_ID = -1
 
 
+class TreeArray:
```
Comment (morgendave) [inline vllm/v1/spec_decode/eagle.py:21]: Expanding in real time is going to be very costly. Dynamic tree in actual production could be less effective

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -233,6 +466,366 @@ def propose(
...
+                self.hidden_states[:num_process_nodes] = batch_hidden_states
+                forward_hidden_states = self.hidden_states
+            else:
+                forward_hidden_states = batch_hidden_states
+
+            # Run model forward pass to get next token predictions
+            with set_forward_context(attn_metadata,
+                                     self.vllm_config,
+                                     num_tokens=input_batch_size):
+                last_hidden_states, output_hidden_states = self.model(
```
Comment (morgendave) [inline vllm/v1/spec_decode/eagle.py:726]: Curious in ROPE kernel do we already take into consideration that positions can be customized?

File: vllm/v1/spec_decode/eagle.py (hunk 1 above)
Comment (wwl2755) [reply vllm/v1/spec_decode/eagle.py:21]: Yes. the sampling logic should be changed also. It is not included yet. You can see a tracker of the progress in: https://docs.google.com/document/d/1mMoSicPPMMzaE_T5Zk2SnTderw1OXRUs2T16JxfVGCQ/edit?usp=sharing

And IMO, from static tree to dynamic tree, it won't introduce much difference (select all/top-k to expand & rerank logic). The major differene are from the tree structure comparing with the chain draft. But I'm open to community's opinions on which should we target on first.

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
Comment (wwl2755) [reply vllm/v1/spec_decode/eagle.py:726]: Thanks for bringing this up! IIUC, in the `rotary_embedding.py`, we could pass an `offsets` to the forward function. 

We have to custom the logic since different path have been mixed together and I would categorize it in "Attention metadata & attention mask" in the tracker. For now, it is only a place-holder.

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -271,6 +500,362 @@ def propose(
...
+                self.hidden_states[:num_process_nodes] = batch_hidden_states
+                forward_hidden_states = self.hidden_states
+            else:
+                forward_hidden_states = batch_hidden_states
+
+            # Run model forward pass to get next token predictions
+            with set_forward_context(tree_per_layer_attn_metadata,
+                                     self.vllm_config,
+                                     num_tokens=input_batch_size):
+                last_hidden_states, output_hidden_states = self.model(
```
Comment (liunianxuxie) [inline vllm/v1/spec_decode/eagle.py:756]: I tried to use the tree_draft_propsoe you provided, but I got error like this
image

**Comment by wwl2755** [comment]
Close because tree attention was supported in https://github.com/vllm-project/vllm/pull/20401
